File: src/stores/history.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { invoke } from '@tauri-apps/api/core'

export interface HistoryEntry {
    id: string
    serviceId: string | null
    endpointId: string | null
    method: string
    url: string
    requestHeaders: { name: string; value: string }[]
    requestBody: string
    responseStatus: number
    responseStatusText: string
    responseHeaders: { name: string; value: string }[]
    responseBody: string
    timeElapsed: number
    size: number
    createdAt: string
}

export const useHistoryStore = defineStore('history', () => {
    const history = ref<HistoryEntry[]>([])
    const isLoading = ref(false)
    const error = ref<string | null>(null)

    async function fetchHistory(limit: number = 50, offset: number = 0) {
        isLoading.value = true
        try {
            const data = await invoke<HistoryEntry[]>('get_history', { limit, offset })
            history.value = data
        } catch (err: any) {
            error.value = err.toString()
            console.error('Failed to fetch history:', err)
        } finally {
            isLoading.value = false
        }
    }

    async function clearHistory() {
        isLoading.value = true
        try {
            await invoke('clear_history')
            history.value = []
        } catch (err: any) {
            error.value = err.toString()
            console.error('Failed to clear history:', err)
        } finally {
            isLoading.value = false
        }
    }

    return {
        history,
        isLoading,
        error,
        fetchHistory,
        clearHistory
    }
})

File: src-tauri/src/tests/unit/history.rs
use crate::history::HistoryService;
use crate::types::HistoryEntry;
use rusqlite::Connection;

#[test]
fn test_history_lifecycle() {
    let conn = Connection::open_in_memory().unwrap();
    let service = HistoryService::new(conn);

    // Init DB
    assert!(service.init().is_ok());

    // Save entry
    let entry = HistoryEntry {
        id: "h1".to_string(),
        service_id: Some("s1".to_string()),
        endpoint_id: Some("e1".to_string()),
        method: "GET".to_string(),
        url: "/test".to_string(),
        request_headers: vec![],
        request_body: "".to_string(),
        response_status: 200,
        response_status_text: "OK".to_string(),
        response_headers: vec![],
        response_body: "body".to_string(),
        time_elapsed: 10,
        size: 4,
        created_at: "2023-01-01T00:00:00Z".to_string(),
    };

    assert!(service.save(entry.clone()).is_ok());

    // Get history
    let history = service.get_history(10, 0).unwrap();
    assert_eq!(history.len(), 1);
    assert_eq!(history[0].id, "h1");

    // Clear history
    assert!(service.clear().is_ok());
    let history = service.get_history(10, 0).unwrap();
    assert_eq!(history.len(), 0);
}
